lib/services/spiffe_federations.go

```diff
@@ -0,0 +1,126 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package services
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/spiffe/go-spiffe/v2/bundle/spiffebundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+
+	machineidv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/machineid/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// SPIFFEFederations is an interface over the SPIFFEFederations service. This
+// interface may also be implemented by a client to allow remote and local
+// consumers to access the resource in a similar way.
+type SPIFFEFederations interface {
+	// GetSPIFFEFederation gets a SPIFFE Federation by name.
+	GetSPIFFEFederation(
+		ctx context.Context, name string,
+	) (*machineidv1.SPIFFEFederation, error)
+	// ListSPIFFEFederations lists all SPIFFE Federations using Google style
+	// pagination.
+	ListSPIFFEFederations(
+		ctx context.Context, pageSize int, lastToken string,
+	) ([]*machineidv1.SPIFFEFederation, string, error)
+	// CreateSPIFFEFederation creates a new SPIFFE Federation.
+	CreateSPIFFEFederation(
+		ctx context.Context, spiffeFederation *machineidv1.SPIFFEFederation,
+	) (*machineidv1.SPIFFEFederation, error)
+	// DeleteSPIFFEFederation deletes a SPIFFE Federation by name.
+	DeleteSPIFFEFederation(ctx context.Context, name string) error
+}
+
+// MarshalSPIFFEFederation marshals the SPIFFEFederation object into a JSON byte
+// array.
+func MarshalSPIFFEFederation(object *machineidv1.SPIFFEFederation, opts ...MarshalOption) ([]byte, error) {
+	object.Version = types.V1
```

Comment:
should we reset the value here? If we ever change to v2 but using the same message, this field will be problematic

Reply:
Hm, I don't really mind, I could just enforce that the Version and Kind is set to something within ValidateSPIFFEFederation.

Reply:
a610cbb055